lib/reversetunnel/agent.go

```diff
@@ -27,444 +27,331 @@ import (
+	AgentInitial AgentState = "initial"
+	// AgentConnected is the state of an agent when is successfully connects
+	// to a server and sends its first heartbeat.
+	AgentConnected AgentState = "connected"
+	// AgentClosed is the state of an agent when the connection and all other
+	// resources are cleaned up.
+	AgentClosed AgentState = "closed"
 )
 
-// AgentConfig holds configuration for agent
-type AgentConfig struct {
-	// Addr is target address to dial
-	Addr utils.NetAddr
-	// ClusterName is the name of the cluster the tunnel is connected to. When the
-	// agent is running in a proxy, it's the name of the remote cluster, when the
-	// agent is running in a node, it's the name of the local cluster.
-	ClusterName string
-	// Signers contains authentication signer
-	Signer ssh.Signer
-	// Client is a client to the local auth servers
-	Client auth.ClientI
-	// AccessPoint is a caching access point to the local auth servers
-	AccessPoint auth.AccessCache
-	// Context is a parent context
-	Context context.Context
-	// Username is the name of this client used to authenticate on SSH
-	Username string
-	// Clock is a clock passed in tests, if not set wall clock
-	// will be used
-	Clock clockwork.Clock
-	// EventsC is an optional events channel, used for testing purposes
-	EventsC chan string
-	// KubeDialAddr is a dial address for kubernetes proxy
-	KubeDialAddr utils.NetAddr
-	// Server is either an SSH or application server. It can handle a connection
-	// (perform handshake and handle request).
-	Server ServerHandler
-	// ReverseTunnelServer holds all reverse tunnel connections.
-	ReverseTunnelServer Server
-	// LocalClusterName is the name of the cluster this agent is running in.
-	LocalClusterName string
-	// Component is the teleport component that this agent runs in.
-	// It's important for routing incoming requests for local services (like an
-	// IoT node or kubernetes service).
-	Component string
-	// Tracker tracks proxy
-	Tracker *track.Tracker
-	// Lease manages gossip and exclusive claims.  Lease may be nil
-	// when used in the context of tests.
-	Lease track.Lease
-	// Log optionally specifies the logger
-	Log log.FieldLogger
-	// FIPS indicates if Teleport was started in FIPS mode.
-	FIPS bool
-	// reverseTunnelDetails cacheable details about the Addr endpoint used to reduce proxy ping calls in order to prevent
-	// proxy endpoint stagnation where even numbers of proxy are hidden behind RoundRobbin Load Balancer.
-	// For instance in a situation where only two proxies [A, B] are configured behind RoundRobbin Load Balancer
-	// due to sequential Ping, Dial method order and sequential backend picking by RoundRobbing Load Balancer
-	// the Ping call will always reach Proxy A and the Dial call will always be forwarded by the LB to Proxy B.
-	reverseTunnelDetails *reverseTunnelDetails
+// AgentStateCallback is called when an agent's state changes.
+type AgentStateCallback func(*Agent)
```

Comment:
Should this really pass the entire agent into the callback? Would it make more sense to provide just the new state instead?

Reply:
I suppose we could but part of the reason its done this way is so the agent store can have a reference to the agent state to remove it from the agent store. The callback could be something like...
```
func (p *AgentPool) GetAgentStateCallback(a Agent) AgentStateCallback {
    return func(s AgentState) {
        if s == closed {
            p.active.remove(a)
        }
    }
}
```
but this seems a little more complicated to me.

Reply:
Ended up going with the above [4305244](https://github.com/gravitational/teleport/pull/10444/commits/4305244b656ca080575a7049b3f72a547b8c134f)